tachyon/math/polynomials/univariate/radix2_evaluation_domain.h

```diff
@@ -267,6 +345,156 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree> {
     }
   }
 
+  // This can be used as the first half of a parallelized butterfly network.
+  CONSTEXPR_IF_NOT_OPENMP void ParDitLayer(RowMajorMatrix<F>& mat, bool inv) {
+    size_t cols = static_cast<size_t>(mat.cols());
+    CHECK(base::bits::IsPowerOfTwo(this->size_));
+    size_t chunk_rows = 1 << mid_;
+
+    // max block size: 2^|mid_|
+    OPENMP_PARALLEL_FOR(size_t block_start = 0; block_start < this->size_;
+                        block_start += chunk_rows) {
+      size_t cur_chunk_rows = std::min(chunk_rows, this->size_ - block_start);
+      Eigen::Block<RowMajorMatrix<F>> submat =
+          mat.block(block_start, 0, cur_chunk_rows, cols);
+      for (size_t layer = 0; layer < mid_; ++layer) {
+        DitLayer(submat, layer, inv);
+      }
+    }
+  }
+
+  // This can be used as the second half of a parallelized butterfly network.
+  CONSTEXPR_IF_NOT_OPENMP void ParDitLayerRev(
+      RowMajorMatrix<F>& mat, const std::vector<F>& twiddles_rev) {
+    size_t cols = static_cast<size_t>(mat.cols());
+    CHECK(base::bits::IsPowerOfTwo(this->size_));
```

Comment:
Isn't size always power of 2?

Reply:
That's a really good point - especially since ::Create sets the size to be a power of two. I'll have to check `mat.rows()` explicitly then. Thanks for catching that